src/move/research/prover-examples/sources/value_holder.move

```diff
@@ -0,0 +1,129 @@
+module prover_examples::value_holder {
+    use std::signer;
+
+    const DEFAULT_VALUE: u8 = 0;
+
+    struct ValueHolder has key {
+        value: u8
+    }
+
+    struct ValueHolderManifest has key {
+        /// The addresses of all accounts that have a ValueHolder.
+        account_addresses: vector<address>,
+        /// The value that all ValueHolders should have.
+        value: u8
+    }
+
+    spec schema Initialized {
+        requires exists<ValueHolderManifest>(@prover_examples);
+    }
+
+    spec module {
+        apply Initialized to * except init_module;
+    }
+
+    /// Ensure that on the creation of a value holder, the account address is originally absent
+    /// from the manifest then gets added, and that the value in the value holder is the same as
+    /// the value in the manifest.
+    invariant update[suspendable] forall account_address: address where !old(
+        exists<ValueHolder>(account_address)
+    ) && exists<ValueHolder>(account_address):
+        contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        )
+            && old(
+                !contains(
+                    global<ValueHolderManifest>(@prover_examples).account_addresses,
+                    account_address
+                )
+            )
+            && global<ValueHolder>(account_address).value
+                == global<ValueHolderManifest>(@prover_examples).value;
+
+    /// Ensure that if an account address is originally absent from the manifest but is then
+    /// added, a value holder is created at the corresponding address with the correct value.
+    invariant update[suspendable] forall account_address: address where !old(
+        contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        )
+    ) && contains(
+        global<ValueHolderManifest>(@prover_examples).account_addresses,
+        account_address
+    ):
+        exists<ValueHolder>(account_address)
+            && old(!exists<ValueHolder>(account_address))
+            && global<ValueHolder>(account_address).value
+                == global<ValueHolderManifest>(@prover_examples).value;
+
+    /// Ensure that if the value in the manifest is updated, all value holders are updated.
+    invariant update[suspendable] forall new_value: u8, account_address: address where old(
+        exists<ValueHolderManifest>(@prover_examples)
+    ) && old(exists<ValueHolder>(account_address)):
+        global<ValueHolder>(account_address).value == new_value;
```

Comment:
@rahxephon89 in 89487af I added this suspendable invariant as well as the `update_value` function, however I am getting `error: global memory invariant does not hold`. I'm trying to make sense of the errors but unfortunately I find them a bit cryptic:

## First error

```sh
error: global memory invariant does not hold
   ┌─ /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:61:5
   │  
61 │ ╭     invariant update[suspendable] forall new_value: u8, account_address: address where old(
62 │ │         exists<ValueHolderManifest>(@prover_examples)
63 │ │     ) && old(exists<ValueHolder>(account_address)):
64 │ │         global<ValueHolder>(account_address).value == new_value;
   │ ╰────────────────────────────────────────────────────────────────^
   │  
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:66: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:105: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:18
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:106: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:110: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:119: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:66: init_value_holder
   =         account = signer{0xabd}
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:67: init_value_holder
   =         value_holder_manifest_ref_mut =
   =           &value_holder.ValueHolderManifest{
   =             account_addresses =
   =               vector{
   =                 (size): 6335,
   =                 3054: 0xabe,
   =                 4018: 0xa73,
   =                 5143: 0xabc,
   =                 default: 0x0},
   =             value = 35u8}
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:70: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:68: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:72: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:73: init_value_holder
   =     at /Users/yours-truly/.move/https___github_com_aptos-labs_aptos-framework_git_mainnet/aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:12: address_of
   =         s = signer{0xabd}
   =     at /Users/yours-truly/.move/https___github_com_aptos-labs_aptos-framework_git_mainnet/aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:13: address_of
   =         result = 0xabd
   =     at /Users/yours-truly/.move/https___github_com_aptos-labs_aptos-framework_git_mainnet/aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:14: address_of
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:72: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:75: init_value_holder
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:113: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:114: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:115: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:120: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:124: init_value_holder (spec)
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:28
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:46
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:61
```

1. How can I interpret the `vector{(size):...` part? 
2. Are the subsequent calls showing a failing counterexample?

## Second failure

```sh
error: global memory invariant does not hold
   ┌─ /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:61:5
   │  
61 │ ╭     invariant update[suspendable] forall new_value: u8, account_address: address where old(
62 │ │         exists<ValueHolderManifest>(@prover_examples)
63 │ │     ) && old(exists<ValueHolder>(account_address)):
64 │ │         global<ValueHolder>(account_address).value == new_value;
   │ ╰────────────────────────────────────────────────────────────────^
   │  
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:77: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:18
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:77: update_value
   =         new_value = 41u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:78: update_value
   =         value_holder_manifest_ref_mut =
   =           &value_holder.ValueHolderManifest{
   =             account_addresses =
   =               vector{(size): 29359, 29358: 0x7c0, default: 0x1},
   =             value = 190u8}
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:79: update_value
   =         $t3 = &190u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:80: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:77: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:80: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:28
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:46
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:61
```

1. How can I interpret the `vector{(size):...` part? 
2. Are the subsequent calls showing a failing counterexample?

Reply:
1) ```
          &value_holder.ValueHolderManifest{
   =             account_addresses =
   =               vector{
   =                 (size): 6335, // size of the vector 
   =                 3054: 0xabe, // address is 0xabe at index 3054, 
   =                 4018: 0xa73, // address is 0xa73 at index 4018, 
   =                 5143: 0xabc, // address is 0xabc at index 3054, 
   =                 default: 0x0}, // at other places, address are all 0x0
   =             value = 35u8} // ValueHolderManifest.value is 35
```
2) Yes, it shows a counter example trace, if you use `aptos move prove -t`, it should give back a more detailed trace.

Reply:
@rahxephon89 per your above and as of b239b93 I've added several invariants to augment the one that was originally causing errors

However now I am getting a trace that I am unable to make sense of:

```sh
error: global memory invariant does not hold
   ┌─ /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:53:5
   │  
53 │ ╭     invariant [suspendable] forall account_address: address where exists<ValueHolder>(
54 │ │         account_address
55 │ │     ):
56 │ │         exists<ValueHolderManifest>(@prover_examples)
   · │
61 │ │             && global<ValueHolder>(account_address).value
62 │ │                 == global<ValueHolderManifest>(@prover_examples).value;
   │ ╰───────────────────────────────────────────────────────────────────────^
   │  
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:128: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:40
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:128: update_value
   =         new_value = 32u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:129: update_value
   =         value_holder_manifest_ref_mut =
   =           &value_holder.ValueHolderManifest{
   =             account_addresses = vector{0xabe, 0xabd},
   =             value = 33u8}
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:130: update_value
   =         $t3 = &33u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:131: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:19
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:128: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:131: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:48
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:53
```

It appears that the issue is related to a new invariant, namely this one: https://github.com/econia-labs/econia-v5/blob/b239b931000a130c15c0ddb52aef553aa769b8a4/src/move/research/prover-examples/sources/value_holder.move#L51-L62

Can you explain what the issue might be in this case?

Reply:
Noting per https://github.com/econia-labs/econia-v5/pull/59#discussion_r1899842840 that resolution may require a loop invariant

Reply:
The loop invariant cannot be proved because `update_value` uses the line function `for_each_ref`. You could try to use a while loop and add invariants there.